### DISCUSSION TITLE
Polyphase vs linear

### DISCUSSION THREAD

**Discussion by larspalo** [GrandOrgue/grandorgue#349] [General]
Originally posted 2020-09-08 06:06:13.949000 by videophile in GrandOrgue discussion forum on SourceForge<br><br>I cannot hear any difference between linear and polyphase resampling. Since I use a 48 kHz sample set on a 48 kHz output, should I conclude that, in this case, no resampling occurs in GO, thus giving the same result whatever option is selected? Or is there still some resampling involved for internal processing, but for any reason (my audio system, my hears, the keys and stops I used for testing etc.) I don&#039;t hear it?

**Comment by larspalo** [comment]
Originally posted 2020-09-08 08:31:32.368000 by [Lars Palo](https://github.com/larspalo) in GrandOrgue discussion forum on SourceForge<br><br>There is a difference between polyphase and linear, but if you enable
lossless compression there will be some cases where polyphase will fallback
to linear interpolation. In such cases you&#039;ll obviously not hear any
difference.

If you listen very carefully at how the high treble and low bass is treated
you should be able to hear the difference in interpolation methods when not
using lossless compression.

&gt;
<br><br>Attachments in original post:<br>https://sourceforge.net/p/ourorgan/discussion/962124/thread/3b1d84f2b6/9d7c/2799/attachment/alternate<br>

**Comment by larspalo** [comment]
Originally posted 2020-09-08 10:39:21.718000 by videophile in GrandOrgue discussion forum on SourceForge<br><br>Yes, if I select lossless compression, GrandOrgue automatically reverts back to linear, I have seen that.

But one point remains unclear: if I use a 48 kHz sample set and a 48 kHz audio output, should I care?

I&#039;d really like the answer to be &quot;no&quot;, because due to my hardware limitation, I have to use lossless compression...

**Reply by larspalo** [reply]
Originally posted 2020-09-08 11:06:18.741000 by [Lars Palo](https://github.com/larspalo) in GrandOrgue discussion forum on SourceForge<br><br>&gt; Yes, if I select lossless compression, GrandOrgue automatically reverts
&gt; back to linear, I have seen that.
&gt;
&gt; But one point remains unclear: if I use a 48 kHz sample set and a 48 kHz
&gt; audio output, should I care?
&gt;
&gt; I&#039;d really like the answer to be &quot;no&quot;, because due to my hardware
&gt; limitation, I have to use lossless compression...
&gt;
No, you don&#039;t need to care. There might be some resampling going on anyway
due to tuning and re-tuning, but you don&#039;t need to be alarmed that you will
be using linear interpolation, just live with it.

In my own experience, linear is actually better in the bass region and
polyphase is better in the treble. You pick what you feel is sounding
better for your usage. Linear outperform polyphase when it comes to
available polyphony, but again it&#039;s your choice...

&gt;
<br><br>Attachments in original post:<br>https://sourceforge.net/p/ourorgan/discussion/962124/thread/3b1d84f2b6/44ab/8ec5/attachment/alternate<br>

**Reply by larspalo** [reply]
Originally posted 2020-09-08 11:16:13.136000 by [Nick Appleton](https://github.com/nickappleton) in GrandOrgue discussion forum on SourceForge<br><br>Just to chime in here (as I think I added both of these re-samplers to GrandOrgue back in the day). Here are some facts regarding the re-samplers in GrandOrgue:

1. Resampling in GrandOrgue occurs when ever a pipe sample needs to be re-pitched. This can happen in a variety of situations:

* Choosing a different temperament.
* Changing the fundamental pitch of the instrument.
* Samples that have been shipped without being tuned.

2. The polyphase re-sampler will exhibit a slight low-pass characteristic. From memory, it’s fairly flat up to about 16-18 kHz - but after that, it begins to roll-off quite steeply. This is the price that must be paid for removing aliasing and imaging artefacts. The linear re-sampler removes close to no aliasing or imaging. Aliasing and imaging are the names given to the undesirable effects that occur during resampling - specifically, the input signal gets upsampled (sample rate is increased), filtered and downsampled (sampling rate decreased). Imaging is produced by the upsampling operation. Aliasing is produced by the downsampling operation IFF the signal before downsampling has frequency components that cannot be represented after the downsampling. The filter’s job is to remove the images and ensure that there will not be aliases. Both the polyphase and the linear interpolation can be modelled in this way and the difference is the filter. I wrote a blog on this a while back: https://www.appletonaudio.com/blog/2015/real-time-re-sampling-and-linear-interpolation/

3. The audible impact of resampling is more pronounced when there is more resampling. If only a small amount of resampling is occurring it will be hard to hear issues. Things become much easier to hear specifically when the pitch of a pipe is *increased* (this is much worse than decreasing - at least for organ samples). You will almost certainly hear no difference for low frequency flue pipes. You will hear terrible artefacts in higher-pitched mixtures - again, pronounced if the pitch is being increased.

The short version of this is:
* Resampling almost always happens - regardless of the sample rate.
* Increasing the pitch of samples is almost always worse than decreasing the pitch.
* Bass pipes aren’t going to care about the kind of interpolation used.
* Mixtures or any high-frequency pipes are the most impacted (play top notes of mixtures and you should hear frequencies which are not harmonic anymore - it’ll sound like there is something else playing that is wildly out of tune).

Nick


&gt; On 8 Sep 2020, at 8:39 pm, Videophile &lt;videophile@users.sourceforge.net&gt; wrote:
&gt; 
&gt; Yes, if I select lossless compression, GrandOrgue automatically reverts back to linear, I have seen that.
&gt; 
&gt; But one point remains unclear: if I use a 48 kHz sample set and a 48 kHz audio output, should I care?
&gt; 
&gt; I&#039;d really like the answer to be &quot;no&quot;, because due to my hardware limitation, I have to use lossless compression... 
&gt; 
&gt; 
&gt; ---
&gt; 
&gt; [Polyphase vs linear ](https://sourceforge.net/p/ourorgan/discussion/962124/thread/3b1d84f2b6/?limit=25#44ab)
&gt; 
&gt; 
&gt; ---
&gt; 
&gt; Sent from sourceforge.net because you indicated interest in &lt;https://sourceforge.net/p/ourorgan/discussion/962124/&gt;
&gt; 
&gt; 
&gt; 
&gt; To unsubscribe from further messages, please visit &lt;https://sourceforge.net/auth/subscriptions/&gt;

**Comment by larspalo** [comment]
Originally posted 2020-09-08 11:16:21.130000 by videophile in GrandOrgue discussion forum on SourceForge<br><br>That&#039;s good news, thank you!

Out of curiosity, I will check my .organ file to see if there is tuning going on. 

Edit: the above answer was to Lars Palo.

@nickappleton Appleton: I use the original temperament, I didn&#039;t change the fundamental pitch, and I hope that the Friesach organ I use was in tune or at least only few pipes were off (I am not at home for the moment, but as soon as I can I will search in the .organ file for tuned pipes). In this case, resampling should almost *never* occur, right?

**Reply by larspalo** [reply]
Originally posted 2020-09-08 22:30:13.008000 by [Nick Appleton](https://github.com/nickappleton) in GrandOrgue discussion forum on SourceForge<br><br>First, sorry for the layout of the post on the forum, I respond to these via email and didn’t expect all this odd formatting.

&gt; Out of curiosity, I will check my .organ file to see if there is tuning going on.
&gt; @nickappleton Appleton: I use the original temperament, I didn&#039;t change the fundamental pitch, and I hope that the Friesach organ I use was in tune or at least only few pipes were off (I am not at home for the moment, but as soon as I can I will search in the .organ file for tuned pipes). In this case, resampling should almost never occur, right?

The .organ file might not tell you very much - but Lars will be more of an expert here. Much has changed in the .organ files since I last worked on GrandOrgue. My opinion here is that all tuning information should be stored in the samples themselves - there should be no explicit per-pipe tuning information in the .organ file. I have some vague recollection that support for doing this was added at some point in the past, but I would hope that this functionality is not being used with Piotr’s sample sets.

It’s unlikely that the samples will be in tune. Tuning the samples prior to releasing a sample set is really a legacy that the original Hauptwerk v1 required back in the days were a computer was not powerful enough to do resampling on thousands of pipe samples independently. Samples don’t need to be in tune anymore because:

1. Changing temperaments or retuning an older church organ (unlikely to be at concert pitch) to concert pitch is going to require resampling anyway.
2. Sometimes people like to hear the instrument as it sounded with no re-samplers (although, this is actually a bit silly because organ pipes “pull” each other into tune - particularly as their pitch increases i.e. when tuning a 2 ft stop, you would normally also have the 4 ft engaged and also an 8 ft from a particular note. When tuning mixtures you would also normally have the 4 ft and 8 ft engaged. You can try this on a real instrument that is in tune, play pipes on the mixture with chorus stops engaged and then turn the chorus stops off and sometimes the mixture will start to beat. This is also the reason that celeste stops are normally placed on the other end of a soundboard to the string it is meant to beat with: put them too close to each other and they will pull each other and stop beating! It is also the reason that I think that randomising pipe speech is not actually a realistic feature - pipes always will end up playing with the same phase on a real instrument. I’ll stop ranting now.
3. Doing the resampling is work that doesn’t need to happen anymore because the virtual organ software contains re-samplers :)

Nick


&gt; On 8 Sep 2020, at 9:16 pm, Videophile &lt;videophile@users.sourceforge.net&gt; wrote:
&gt; 
&gt; That&#039;s good news, thank you!
&gt; 
&gt; Out of curiosity, I will check my .organ file to see if there is tuning going on. 
&gt; 
&gt; 
&gt; ---
&gt; 
&gt; [Polyphase vs linear ](https://sourceforge.net/p/ourorgan/discussion/962124/thread/3b1d84f2b6/?limit=25#25ca)
&gt; 
&gt; 
&gt; ---
&gt; 
&gt; Sent from sourceforge.net because you indicated interest in &lt;https://sourceforge.net/p/ourorgan/discussion/962124/&gt;
&gt; 
&gt; 
&gt; 
&gt; To unsubscribe from further messages, please visit &lt;https://sourceforge.net/auth/subscriptions/&gt;

**Reply by larspalo** [reply]
Originally posted 2020-09-09 05:43:29.673000 by [Lars Palo](https://github.com/larspalo) in GrandOrgue discussion forum on SourceForge<br><br>&gt; The .organ file might not tell you very much - but Lars will be more of an
&gt; expert here. Much has changed in the .organ files since I last worked on
&gt; GrandOrgue. My opinion here is that all tuning information should be stored
&gt; in the samples themselves - there should be no explicit per-pipe tuning
&gt; information in the .organ file. I have some vague recollection that support
&gt; for doing this was added at some point in the past, but I would hope that
&gt; this functionality is not being used with Piotr’s sample sets.
&gt;
Pitch info can be embedded in the .wav files, but the exact same info can
be specified in the ODF too if the samples don&#039;t have it. Additionally,
there can be different kinds of re-tuning info present that will adjust the
tuning depending on whether any special temperament is active or not
(PitchTuning and PitchCorrection). User settings can also influence the
pitch down to per pipe level.

The reasons for having possible pitch info in the ODF are many. The
embedded pitch info (together with specified harmonica number) is used only
if a specific temperament is chosen - ie not for original temperament. In
some cases one wants to extend the range of the manual/rank without
additional samples. In Bureå Choir organ I actually only samples 4 notes
per octave and interpolated the rest by retuning. Then another reason is to
correct obvious mistakes in the original temperament that you don&#039;t want
the user to have to fix themselves (either by selecting a specific
temperament or supplying the right correction in cents for the pipe).

Then there can be a historic organ at a different pitch, in which case you
might not want a selection of temperament to also change the overall pitch
(which it will do by default). In such cases it&#039;s necessary to add extra
info to prevent that. The possibility for the user to change overall pitch
of course remains in more than one way...

&gt;
<br><br>Attachments in original post:<br>https://sourceforge.net/p/ourorgan/discussion/962124/thread/3b1d84f2b6/25ca/cae0/b648/attachment/alternate<br>

**Reply by larspalo** [reply]
Originally posted 2020-10-19 03:25:26.918000 by dhanaseelanprav in GrandOrgue discussion forum on SourceForge<br><br>Does this sample work with Grandorgue software?

**Reply by nickappleton** [reply]
I'm confused. What happened here? Did content get copied across from an old sourceforge discussion or is there something new in this?

**Reply by larspalo** [reply]
@nickappleton Yes, this is a migration of the old discussion forum from Sourceforge. However, if you have time, still we don't have a working polyphase when lossless compression is selected! ;-)

**Reply by nickappleton** [reply]
This would probably make for a good issue ticket so it does not get forgotten.

As for time to address it: working full-time in lockdown with a 4 year old and a 6 month old leaves me with very little time to do anything anymore :)

**Reply by larspalo** [reply]
I'll add it for the future.

Quite understandable! Enjoy the time with the kids! I've got five of them, the oldest 14 years old and the youngest 5 years old.
